### ISSUE TITLE
proposal: optimize the memory usage when EnableMemory = true for StructLogger 

### ISSUE THREAD

**Issue by lispc** [ethereum/go-ethereum#24873]
# Rationale

Current code does [a full memory snapshot for every execution step](https://github.com/ethereum/go-ethereum/blob/440c9fcf75d9d5383b72646a65d5e21fa7ab6a26/eth/tracers/logger/logger.go#L167).

While for storage, [it only snapshots storage when the opcode accesses storage](https://github.com/ethereum/go-ethereum/blob/440c9fcf75d9d5383b72646a65d5e21fa7ab6a26/eth/tracers/logger/logger.go#L183) (SLOAD & SSTORE here).

So my proposal: only snapshot the full memory when the opcode updates memory. If not, we directly set memory in StructLog to empty slice.   Since evm cannot really free memory, when a StructLog of step contains empty memory, we know this means the step does not update memory, memory stays the same like previous step.

Reason for this proposal: When EnableMemory = true, if we trace a mainnet block, it costs several minutes but still failed even in a server running geth with 64GiB memory. If this proposal is applied, much fewer steps are needed to take a full snapshot of memory, so tracing will cost less time and memory.

backwards compatibility: it will break compatibility. But codes using this feature will be pretty easy to adapt the new interface.

# Implementation

Are you willing to implement this feature?: Yes

**Comment by lispc** [comment]
another approach is to add a flag named "MemoryTraceMode".  

MemoryTraceMode == 0: snapshot full memory every step; 
MemoryTraceMode == 1: snapshot full memory when updated;
MemoryTraceMode == 2: snapshot memory diff

**Comment by s1na** [comment]
I agree this would be nice as traces are usually pretty heavy. We can deprecate `EnableMemory` and set `MemoryTraceMode` by default to a value that implies memory tracing is disabled for backwards-compatibility.

**Comment by s1na** [comment]
Can you please try this [patch](https://github.com/s1na/go-ethereum/tree/logger/mem-dedup) on the mainnet block you mentioned and see if it makes a big difference?

**Comment by s1na** [comment]
This PR might also be of interest: https://github.com/ethereum/go-ethereum/pull/24895

**Comment by lispc** [comment]
I analyzed(traced with mem trace enabled) every tx inside the mentioned block. There are 94 txs in total, 93 of them can be traced successfully(the biggest trace is of ~300MiB size, summing all is about ~900MiB), [the failed one(always makes geth OOM) is an Arbitrum Sequencer tx](https://etherscan.io/tx/0x0ef22f22aa244bdbf2e6b49b6b36f71717f9d18c2a3d7b24be52ff26abd1a220), having calldata of ~100KiB. In the Arbitrum contract, the calldata are copied to memory and hashed. So snapshots of mem for every step will be HUGE..

**Comment by holiman** [comment]
Triage discussion: an alternate variant would be to

> If memory changes between step N and N+1, then output the full memory at N+1. 


This means that at any scope-changing operation (revert, return, call-variants) at N, step N+1 would output the memory (unless it was empty in the old and new scope). And also output it at MSTORE/MSTORE8 ops. 

Such format change would bring down the output size __a lot__, but still make it pretty easy to build a non-intelligent traceviewer. I want to avoid forcing a traceviewer to become a mini-evm.

**Comment by s1na** [comment]
cc @yann300

**Comment by MariusVanDerWijden** [comment]
The StructLogger recently was made streamable in https://github.com/ethereum/go-ethereum/pull/30806 which should make it use way less memory

**Comment by holiman** [comment]
Anything returned over http (at least our http rpc implementation) is inherently un-streamable. So this is still relevant.
